File: Heap/Easy/chinky_and_diamonds.cpp
// https://practice.geeksforgeeks.org/problems/chinky-and-diamonds3340/1#
// { Driver Code Starts

#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends

class Solution {
  public:
    int maxDiamonds(int A[], int N, int K) {
        priority_queue<int> pq;
        
        for(int i = 0; i < N; i++)
            pq.push(A[i]);
        
        int ans = 0;
        for(int i = 0; i < K; i++)
        {
            int top = pq.top();
            // cout << top << endl;
            pq.pop();
            ans += top;
            pq.push(top/2);
        }
        
        
        return ans;
    }
};

// { Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int N,K;
        
        cin>>N>>K;
        int A[N];
        
        for(int i=0; i<N; i++)
            cin>>A[i];

        Solution ob;
        cout << ob.maxDiamonds(A,N,K) << endl;
    }
    return 0;
}  // } Driver Code Ends

File: Heap/Easy/chinky_and_diamonds.py
# https://practice.geeksforgeeks.org/problems/chinky-and-diamonds3340/1#
#User function Template for python3

from heapq import heappop, heappush, heapify
class Solution:
    def maxDiamonds(self, A, N, K):
        pq = []
        heapify(pq)
        
        for i in A:
            heappush(pq, -1 * i)
            
        ans = 0
        while(K):
            ele = heappop(pq)
            ele = (-1 * ele)
            ans += ele
            heappush(pq, -1 * (ele // 2))
            K -= 1    
        
        return ans

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        N,K=map(int,input().split())
        A=list(map(int,input().split()))
        
        ob = Solution()
        print(ob.maxDiamonds(A,N,K))
# } Driver Code Ends